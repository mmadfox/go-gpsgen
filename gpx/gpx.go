package gpx

import (
	"fmt"

	"github.com/mmadfox/go-gpsgen/navigator"
	"github.com/tkrajina/gpxgo/gpx"
)

func Encode(routes []*navigator.Route) ([]byte, error) {
	gpxData := new(gpx.GPX)
	gpxData.Tracks = make([]gpx.GPXTrack, 0)
	gpxData.AuthorName = "go-gpsgen"
	for i := 0; i < len(routes); i++ {
		track := gpx.GPXTrack{
			Segments: make([]gpx.GPXTrackSegment, 0),
		}
		route := routes[i]
		for j := 0; j < route.NumTracks(); j++ {
			trackPoint := make([]gpx.GPXPoint, 0)
			route.EachSegment(j, func(seg *navigator.Segment) {
				trackPoint = append(trackPoint, gpx.GPXPoint{
					Point: gpx.Point{
						Latitude:  seg.PointA().X,
						Longitude: seg.PointA().Y,
					},
				})
				trackPoint = append(trackPoint, gpx.GPXPoint{
					Point: gpx.Point{
						Latitude:  seg.PointB().X,
						Longitude: seg.PointB().Y,
					},
				})
			})
			seq := gpx.GPXTrackSegment{
				Points: trackPoint,
			}
			track.Segments = append(track.Segments, seq)
		}
		gpxData.Tracks = append(gpxData.Tracks, track)
	}
	return gpxData.ToXml(gpx.ToXmlParams{Version: "1.1", Indent: true})
}

func Decode(data []byte) ([]*navigator.Route, error) {
	if len(data) == 0 {
		return nil, fmt.Errorf("gpx: file is empty")
	}
	gpxData, err := gpx.ParseBytes(data)
	if err != nil {
		return nil, err
	}
	routes := make([]*navigator.Route, 0, 4)
	if len(gpxData.Routes) > 0 {
		for i := 0; i < len(gpxData.Routes); i++ {
			points := make([]navigator.Point, 0, len(gpxData.Routes[i].Points))
			for j := 0; j < len(gpxData.Routes[i].Points); j++ {
				points = append(points, navigator.Point{
					X: gpxData.Routes[i].Points[j].Latitude,
					Y: gpxData.Routes[i].Points[j].Longitude,
				})
			}
			route, err := navigator.NewRoute([][]navigator.Point{points})
			if err != nil {
				return nil, err
			}
			routes = append(routes, route)
		}
	}
	if len(gpxData.Tracks) > 0 {
		for i := 0; i < len(gpxData.Tracks); i++ {
			points := make([][]navigator.Point, 0)
			for j := 0; j < len(gpxData.Tracks[i].Segments); j++ {
				segment := make([]navigator.Point, 0, 8)
				for x := 0; x < len(gpxData.Tracks[i].Segments[j].Points); x++ {
					segment = append(segment, navigator.Point{
						X: gpxData.Tracks[i].Segments[j].Points[x].Latitude,
						Y: gpxData.Tracks[i].Segments[j].Points[x].Longitude,
					})
				}
				points = append(points, segment)
			}
			route, err := navigator.NewRoute(points)
			if err != nil {
				return nil, err
			}
			routes = append(routes, route)
		}
	}
	if len(gpxData.Waypoints) > 0 {
		points := make([]navigator.Point, 0, len(gpxData.Waypoints))
		for i := 0; i < len(gpxData.Waypoints); i++ {
			points = append(points, navigator.Point{
				X: gpxData.Waypoints[i].Latitude,
				Y: gpxData.Waypoints[i].Longitude,
			})
		}
		route, err := navigator.NewRoute([][]navigator.Point{points})
		if err != nil {
			return nil, err
		}
		routes = append(routes, route)
	}
	return routes, nil
}
