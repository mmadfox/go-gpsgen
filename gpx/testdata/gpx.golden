<?xml version="1.0" encoding="UTF-8"?>
<gpx  xmlns="http://www.topografix.com/GPX/1/1" version="1.1" creator="https://github.com/tkrajina/gpxgo">
	<metadata>
		<author>
			<name>go-gpsgen</name>
		</author>
	</metadata>
	<trk>
		<trkseg>
			<trkpt lat="30.526812518466187" lon="114.3053712051398"></trkpt>
			<trkpt lat="30.528997477781047" lon="114.29851108401101"></trkpt>
			<trkpt lat="30.528997477781047" lon="114.29851108401101"></trkpt>
			<trkpt lat="30.528997477781047" lon="114.29851108401101"></trkpt>
			<trkpt lat="30.528997477781047" lon="114.29851108401101"></trkpt>
			<trkpt lat="30.528680496119037" lon="114.29341832954645"></trkpt>
			<trkpt lat="30.528680496119037" lon="114.29341832954645"></trkpt>
			<trkpt lat="30.528680496119037" lon="114.29341832954645"></trkpt>
			<trkpt lat="30.528680496119037" lon="114.29341832954645"></trkpt>
			<trkpt lat="30.52812436291228" lon="114.28771999609836"></trkpt>
			<trkpt lat="30.52812436291228" lon="114.28771999609836"></trkpt>
			<trkpt lat="30.52812436291228" lon="114.28771999609836"></trkpt>
			<trkpt lat="30.52812436291228" lon="114.28771999609836"></trkpt>
			<trkpt lat="30.533313934919704" lon="114.27807854525355"></trkpt>
			<trkpt lat="30.533313934919704" lon="114.27807854525355"></trkpt>
			<trkpt lat="30.533313934919704" lon="114.27807854525355"></trkpt>
			<trkpt lat="30.533313934919704" lon="114.27807854525355"></trkpt>
			<trkpt lat="30.539819582793626" lon="114.2677544830193"></trkpt>
			<trkpt lat="30.539819582793626" lon="114.2677544830193"></trkpt>
			<trkpt lat="30.539819582793626" lon="114.2677544830193"></trkpt>
			<trkpt lat="30.539819582793626" lon="114.2677544830193"></trkpt>
			<trkpt lat="30.540593949402663" lon="114.26114290172967"></trkpt>
			<trkpt lat="30.540593949402663" lon="114.26114290172967"></trkpt>
			<trkpt lat="30.540593949402663" lon="114.26114290172967"></trkpt>
			<trkpt lat="30.540593949402663" lon="114.26114290172967"></trkpt>
			<trkpt lat="30.55197076102921" lon="114.26099678731612"></trkpt>
			<trkpt lat="30.55197076102921" lon="114.26099678731612"></trkpt>
			<trkpt lat="30.55197076102921" lon="114.26099678731612"></trkpt>
			<trkpt lat="30.55197076102921" lon="114.26099678731612"></trkpt>
			<trkpt lat="30.557176547041138" lon="114.26050647676828"></trkpt>
			<trkpt lat="30.557176547041138" lon="114.26050647676828"></trkpt>
			<trkpt lat="30.557176547041138" lon="114.26050647676828"></trkpt>
			<trkpt lat="30.557176547041138" lon="114.26050647676828"></trkpt>
			<trkpt lat="30.565791147094203" lon="114.26260674480358"></trkpt>
			<trkpt lat="30.565791147094203" lon="114.26260674480358"></trkpt>
			<trkpt lat="30.565791147094203" lon="114.26260674480358"></trkpt>
			<trkpt lat="30.565791147094203" lon="114.26260674480358"></trkpt>
			<trkpt lat="30.578608614150724" lon="114.26335848475395"></trkpt>
			<trkpt lat="30.578608614150724" lon="114.26335848475395"></trkpt>
			<trkpt lat="30.578608614150724" lon="114.26335848475395"></trkpt>
			<trkpt lat="30.578608614150724" lon="114.26335848475395"></trkpt>
			<trkpt lat="30.583003550274867" lon="114.26161056778528"></trkpt>
			<trkpt lat="30.583003550274867" lon="114.26161056778528"></trkpt>
			<trkpt lat="30.583003550274867" lon="114.26161056778528"></trkpt>
			<trkpt lat="30.583003550274867" lon="114.26161056778528"></trkpt>
			<trkpt lat="30.588136976930087" lon="114.27956491666293"></trkpt>
			<trkpt lat="30.588136976930087" lon="114.27956491666293"></trkpt>
			<trkpt lat="30.588136976930087" lon="114.27956491666293"></trkpt>
			<trkpt lat="30.588136976930087" lon="114.27956491666293"></trkpt>
			<trkpt lat="30.597663532723345" lon="114.28836587615984"></trkpt>
			<trkpt lat="30.597663532723345" lon="114.28836587615984"></trkpt>
			<trkpt lat="30.597663532723345" lon="114.28836587615984"></trkpt>
			<trkpt lat="30.597663532723345" lon="114.28836587615984"></trkpt>
			<trkpt lat="30.605742661662404" lon="114.29699175341466"></trkpt>
			<trkpt lat="30.605742661662404" lon="114.29699175341466"></trkpt>
			<trkpt lat="30.605742661662404" lon="114.29699175341466"></trkpt>
			<trkpt lat="30.605742661662404" lon="114.29699175341466"></trkpt>
			<trkpt lat="30.61756793358738" lon="114.3063371323068"></trkpt>
			<trkpt lat="30.61756793358738" lon="114.3063371323068"></trkpt>
			<trkpt lat="30.61756793358738" lon="114.3063371323068"></trkpt>
			<trkpt lat="30.61756793358738" lon="114.3063371323068"></trkpt>
			<trkpt lat="30.631666925597727" lon="114.31632453637366"></trkpt>
			<trkpt lat="30.631666925597727" lon="114.31632453637366"></trkpt>
			<trkpt lat="30.631666925597727" lon="114.31632453637366"></trkpt>
			<trkpt lat="30.631666925597727" lon="114.31632453637366"></trkpt>
			<trkpt lat="30.63921322609346" lon="114.32176579486497"></trkpt>
			<trkpt lat="30.63921322609346" lon="114.32176579486497"></trkpt>
			<trkpt lat="30.63921322609346" lon="114.32176579486497"></trkpt>
			<trkpt lat="30.63921322609346" lon="114.32176579486497"></trkpt>
			<trkpt lat="30.627390229093507" lon="114.28362121238996"></trkpt>
			<trkpt lat="30.627390229093507" lon="114.28362121238996"></trkpt>
			<trkpt lat="30.627390229093507" lon="114.28362121238996"></trkpt>
			<trkpt lat="30.627390229093507" lon="114.28362121238996"></trkpt>
			<trkpt lat="30.626831821963734" lon="114.28024759376382"></trkpt>
			<trkpt lat="30.626831821963734" lon="114.28024759376382"></trkpt>
			<trkpt lat="30.626831821963734" lon="114.28024759376382"></trkpt>
			<trkpt lat="30.626831821963734" lon="114.28024759376382"></trkpt>
			<trkpt lat="30.613717167827105" lon="114.28059648666806"></trkpt>
			<trkpt lat="30.613717167827105" lon="114.28059648666806"></trkpt>
			<trkpt lat="30.613717167827105" lon="114.28059648666806"></trkpt>
			<trkpt lat="30.613717167827105" lon="114.28059648666806"></trkpt>
			<trkpt lat="30.611518529103265" lon="114.28029184923957"></trkpt>
			<trkpt lat="30.611518529103265" lon="114.28029184923957"></trkpt>
			<trkpt lat="30.611518529103265" lon="114.28029184923957"></trkpt>
			<trkpt lat="30.611518529103265" lon="114.28029184923957"></trkpt>
			<trkpt lat="30.610126760260044" lon="114.2812356800141"></trkpt>
			<trkpt lat="30.610126760260044" lon="114.2812356800141"></trkpt>
			<trkpt lat="30.610126760260044" lon="114.2812356800141"></trkpt>
			<trkpt lat="30.610126760260044" lon="114.2812356800141"></trkpt>
			<trkpt lat="30.6039826842769" lon="114.27626101188014"></trkpt>
			<trkpt lat="30.6039826842769" lon="114.27626101188014"></trkpt>
			<trkpt lat="30.6039826842769" lon="114.27626101188014"></trkpt>
			<trkpt lat="30.6039826842769" lon="114.27626101188014"></trkpt>
			<trkpt lat="30.60027262866558" lon="114.2707216344574"></trkpt>
			<trkpt lat="30.60027262866558" lon="114.2707216344574"></trkpt>
			<trkpt lat="30.60027262866558" lon="114.2707216344574"></trkpt>
			<trkpt lat="30.60027262866558" lon="114.2707216344574"></trkpt>
			<trkpt lat="30.597731517193594" lon="114.26305306099147"></trkpt>
			<trkpt lat="30.597731517193594" lon="114.26305306099147"></trkpt>
			<trkpt lat="30.597731517193594" lon="114.26305306099147"></trkpt>
			<trkpt lat="30.597731517193594" lon="114.26305306099147"></trkpt>
			<trkpt lat="30.59631015724898" lon="114.2581441842604"></trkpt>
			<trkpt lat="30.59631015724898" lon="114.2581441842604"></trkpt>
			<trkpt lat="30.59631015724898" lon="114.2581441842604"></trkpt>
			<trkpt lat="30.59631015724898" lon="114.2581441842604"></trkpt>
			<trkpt lat="30.593438430116905" lon="114.25267083463814"></trkpt>
			<trkpt lat="30.593438430116905" lon="114.25267083463814"></trkpt>
			<trkpt lat="30.593438430116905" lon="114.25267083463814"></trkpt>
			<trkpt lat="30.593438430116905" lon="114.25267083463814"></trkpt>
			<trkpt lat="30.586972034650003" lon="114.24747304440945"></trkpt>
			<trkpt lat="30.586972034650003" lon="114.24747304440945"></trkpt>
			<trkpt lat="30.586972034650003" lon="114.24747304440945"></trkpt>
			<trkpt lat="30.586972034650003" lon="114.24747304440945"></trkpt>
			<trkpt lat="30.581273690184105" lon="114.24744426825725"></trkpt>
			<trkpt lat="30.581273690184105" lon="114.24744426825725"></trkpt>
			<trkpt lat="30.581273690184105" lon="114.24744426825725"></trkpt>
			<trkpt lat="30.581273690184105" lon="114.24744426825725"></trkpt>
			<trkpt lat="30.57562177956096" lon="114.24596651045829"></trkpt>
			<trkpt lat="30.57562177956096" lon="114.24596651045829"></trkpt>
			<trkpt lat="30.57562177956096" lon="114.24596651045829"></trkpt>
			<trkpt lat="30.57562177956096" lon="114.24596651045829"></trkpt>
			<trkpt lat="30.569714157370257" lon="114.24542857767642"></trkpt>
			<trkpt lat="30.569714157370257" lon="114.24542857767642"></trkpt>
			<trkpt lat="30.569714157370257" lon="114.24542857767642"></trkpt>
			<trkpt lat="30.569714157370257" lon="114.24542857767642"></trkpt>
			<trkpt lat="30.56153819385517" lon="114.2395849706893"></trkpt>
			<trkpt lat="30.56153819385517" lon="114.2395849706893"></trkpt>
			<trkpt lat="30.56153819385517" lon="114.2395849706893"></trkpt>
			<trkpt lat="30.56153819385517" lon="114.2395849706893"></trkpt>
			<trkpt lat="30.55866525792476" lon="114.24344839296145"></trkpt>
			<trkpt lat="30.55866525792476" lon="114.24344839296145"></trkpt>
			<trkpt lat="30.55866525792476" lon="114.24344839296145"></trkpt>
			<trkpt lat="30.55866525792476" lon="114.24344839296145"></trkpt>
			<trkpt lat="30.556609921175593" lon="114.24902005215671"></trkpt>
			<trkpt lat="30.556609921175593" lon="114.24902005215671"></trkpt>
			<trkpt lat="30.556609921175593" lon="114.24902005215671"></trkpt>
			<trkpt lat="30.556609921175593" lon="114.24902005215671"></trkpt>
			<trkpt lat="30.554479837710048" lon="114.25603177640284"></trkpt>
			<trkpt lat="30.554479837710048" lon="114.25603177640284"></trkpt>
			<trkpt lat="30.554479837710048" lon="114.25603177640284"></trkpt>
			<trkpt lat="30.554479837710048" lon="114.25603177640284"></trkpt>
			<trkpt lat="30.55639057381194" lon="114.2616778810214"></trkpt>
			<trkpt lat="30.55639057381194" lon="114.2616778810214"></trkpt>
			<trkpt lat="30.55639057381194" lon="114.2616778810214"></trkpt>
			<trkpt lat="30.55639057381194" lon="114.2616778810214"></trkpt>
			<trkpt lat="30.556420974246493" lon="114.26929572236781"></trkpt>
			<trkpt lat="30.556420974246493" lon="114.26929572236781"></trkpt>
			<trkpt lat="30.556420974246493" lon="114.26929572236781"></trkpt>
			<trkpt lat="30.556420974246493" lon="114.26929572236781"></trkpt>
			<trkpt lat="30.55682783403205" lon="114.27461286256727"></trkpt>
			<trkpt lat="30.55682783403205" lon="114.27461286256727"></trkpt>
			<trkpt lat="30.55682783403205" lon="114.27461286256727"></trkpt>
			<trkpt lat="30.55682783403205" lon="114.27461286256727"></trkpt>
			<trkpt lat="30.54773670500522" lon="114.29080071182096"></trkpt>
			<trkpt lat="30.54773670500522" lon="114.29080071182096"></trkpt>
			<trkpt lat="30.54773670500522" lon="114.29080071182096"></trkpt>
			<trkpt lat="30.54773670500522" lon="114.29080071182096"></trkpt>
			<trkpt lat="30.546702266510025" lon="114.29721013990377"></trkpt>
			<trkpt lat="30.546702266510025" lon="114.29721013990377"></trkpt>
			<trkpt lat="30.546702266510025" lon="114.29721013990377"></trkpt>
			<trkpt lat="30.546702266510025" lon="114.29721013990377"></trkpt>
			<trkpt lat="30.542940839597946" lon="114.3014707139979"></trkpt>
			<trkpt lat="30.542940839597946" lon="114.3014707139979"></trkpt>
			<trkpt lat="30.542940839597946" lon="114.3014707139979"></trkpt>
			<trkpt lat="30.542940839597946" lon="114.3014707139979"></trkpt>
			<trkpt lat="30.54270271658028" lon="114.31397023050971"></trkpt>
			<trkpt lat="30.54270271658028" lon="114.31397023050971"></trkpt>
			<trkpt lat="30.54270271658028" lon="114.31397023050971"></trkpt>
			<trkpt lat="30.54270271658028" lon="114.31397023050971"></trkpt>
			<trkpt lat="30.534760458330908" lon="114.32981874488996"></trkpt>
			<trkpt lat="30.534760458330908" lon="114.32981874488996"></trkpt>
			<trkpt lat="30.534760458330908" lon="114.32981874488996"></trkpt>
			<trkpt lat="30.534760458330908" lon="114.32981874488996"></trkpt>
			<trkpt lat="30.52793946059009" lon="114.32618374484628"></trkpt>
			<trkpt lat="30.52793946059009" lon="114.32618374484628"></trkpt>
			<trkpt lat="30.52793946059009" lon="114.32618374484628"></trkpt>
			<trkpt lat="30.52793946059009" lon="114.32618374484628"></trkpt>
			<trkpt lat="30.525000964116842" lon="114.32277720781678"></trkpt>
			<trkpt lat="30.525000964116842" lon="114.32277720781678"></trkpt>
			<trkpt lat="30.525000964116842" lon="114.32277720781678"></trkpt>
			<trkpt lat="30.525000964116842" lon="114.32277720781678"></trkpt>
			<trkpt lat="30.522109039119115" lon="114.31759209439264"></trkpt>
			<trkpt lat="30.522109039119115" lon="114.31759209439264"></trkpt>
			<trkpt lat="30.522109039119115" lon="114.31759209439264"></trkpt>
			<trkpt lat="30.522109039119115" lon="114.31759209439264"></trkpt>
			<trkpt lat="30.51922484999777" lon="114.32635535532978"></trkpt>
		</trkseg>
	</trk>
</gpx>